File: the-daily-starter/pages/api/daily-quote/quote.ts
import { NextApiRequest, NextApiResponse } from "next";
import { getOrCreateDailyQuoteForToday } from "@/lib/firebase/adminUtils/quoteService";

//api endpoint for fetching today's quote from zenquotes or firebase.
export default async function handler(req: NextApiRequest, res: NextApiResponse) {

    if (req.method !== 'GET') {
        return res.status(405).json({ message: 'Method Not Allowed' });
    }

    try {
        const quote = await getOrCreateDailyQuoteForToday();
        return res.status(200).json(quote);
    } catch (err) {
        console.error("Error getting daily quote at api endpoint, " + err);
        return res.status(500).json({ error: "Failed to fetch daily quote" });
    }
}

File: backend/quoteservice/endpoint/quote.go
package QuoteEndpoint

import (
	"context"
	"encoding/json"
	"log"
	"net/http"
	"services/quoteservice"
	"time"
)

func QuoteHandler(svc *quoteservice.Service) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {

		if r.Method != http.MethodGet {
			http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
			return
		}

		ctx, cancel := context.WithTimeout(r.Context(), 5*time.Second)
		defer cancel()

		quote, err := svc.GetOrCreateDailyQuoteForToday(ctx)
		if err != nil {
			log.Printf("Error fetching daily quote: %v", err)
			http.Error(w, "Failed to fetch daily quote", http.StatusInternalServerError)
			return
		}

		w.Header().Set("Content-Type", "application/json")
		if err := json.NewEncoder(w).Encode(quote); err != nil {
			log.Printf("Error encoding response: %v", err)
			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
			return
		}
	}
}
